classes/metadata_fetcher.php

```diff
@@ -0,0 +1,112 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Utility class for fetching IDP metadata.
+ *
+ * @package    auth_saml2
+ * @author     Sam Chaffee
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace auth_saml2;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../../../lib/filelib.php');
+
+/**
+ * Utility class for fetching IDP metadata.
+ *
+ * @package    auth_saml2
+ * @copyright  Copyright (c) 2017 Blackboard Inc. (http://www.blackboard.com)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class metadata_fetcher {
+    /**
+     * @var array
+     */
+    private $curlinfo = [];
+
+    /**
+     * @var string
+     */
+    private $curlerror = '';
+
+    /**
+     * @var int
+     */
+    private $curlerrorno = 0;
+
+    /**
+     * @param $url
+     * @return bool
+     * @throws \coding_exception
+     */
+    public function fetch($url, $curl = null) {
+        if (!$curl instanceof \curl) {
+            $curl = new \curl();
+        }
+        $options = [
+            'CURLOPT_SSL_VERIFYPEER' => true,
```

Comment:
Have you tested this with self signed or with expired ssl? I've run into issues elsewhere with this in both dev env's and prod. Not sure the best way forward, ideally it would be good to keep this on but explicitly catch this type of failure, especially if cron is going to find the issue later on and it was originally ok. Dunno if it's worth a new config boolean.